components/uniset/src/builder.rs

```diff
@@ -0,0 +1,755 @@
+use std::{char, cmp, cmp::Ordering, ops::RangeBounds};
+
+use crate::{uniset::UnicodeSet, utils::deconstruct_range};
+
+/// UnicodeSet builder wrapper
+///
+/// Provides exposure to builder functions and conversion to UnicodeSet
+#[derive(Default)]
+pub struct UnicodeSetBuilder {
+    intervals: Vec<u32>,
+}
+
+impl UnicodeSetBuilder {
+    /// Returns empty UnicodeSetBuilder
+    pub fn new() -> UnicodeSetBuilder {
+        UnicodeSetBuilder { intervals: vec![] }
+    }
+
+    /// Returns a UnicodeSet and consumes the UnicodeSetBuilder
+    pub fn build(self) -> UnicodeSet {
+        UnicodeSet::from_inversion_list(self.intervals).unwrap()
+    }
+
+    /// Adds the range to the end of the builder
+    fn append(&mut self, start: u32, end: u32) {
+        self.intervals.push(start);
+        self.intervals.push(end);
+    }
+
+    /// Adds the range to the beginning of the builder in a single memory move
+    fn prepend(&mut self, start: u32, end: u32) {
+        let mut new = vec![start, end];
+        new.append(&mut self.intervals);
+        self.intervals = new;
+    }
+
+    /// Add the range to the UnicodeSetBuilder
+    ///
+    /// Accomplishes this through binary search for the start and end indices and merges intervals
+    /// in between with inplace memory. Performs `O(1)` operation if adding to end of list, and `O(N)` otherwise,
+    /// where `N` is the number of endpoints.
+    fn add(&mut self, start: u32, end: u32) {
+        if start >= end {
+            return;
+        }
+        if self.intervals.is_empty() {
+            self.append(start, end);
+        } else {
+            let last_ind = self.intervals.len() - 1;
+            if self.intervals[last_ind - 1] <= start {
+                if self.intervals[last_ind] < start {
+                    self.append(start, end);
+                } else {
+                    self.intervals[last_ind] = cmp::max(end, self.intervals[last_ind]);
+                }
+            } else if end <= self.intervals[1] {
+                if end < self.intervals[0] {
+                    self.prepend(start, end);
+                } else {
+                    self.intervals[0] = cmp::min(self.intervals[0], start);
+                }
+            } else {
+                let s = self.intervals.binary_search(&start);
+                let e = self.intervals.binary_search(&end);
+                let mut s_index = s.unwrap_or_else(|x| x);
+                let mut e_index = e.unwrap_or_else(|x| x);
+                let s_even: bool = s_index % 2 == 0;
+                let e_even: bool = e_index % 2 == 0;
+
+                if s_even && e_even && s.is_err() && e.is_err() && s_index == e_index {
+                    self.intervals.splice(s_index..e_index, vec![start, end]);
+                } else {
+                    let s_equals_e = s_index != e_index;
```

Comment:
It would be `s_not_equals_e`

But I would probably just expand this into the only call site on line 78.

Reply:
the reason why this can't be put into the only call site on line 78 is that is that the `if s_even{}` check modifies `s_index`, while we care about the state of the equals before the modification. I can move it out of the parent `if_else` statement, and use it in the aforementioned `if s_index == e_index` check as well to make the usage less confusing

Reply:
I should have included an example of failure in my comment (there's a test case, `add_to_middle_right_overlap`, which documents it):

##### `s_index != e_index inline`

```
# [10, 20, 40, 50] + [30, 40], expected result = [10, 20, 30, 50]
# s_index = 2, e_index = 2, s.is_err(), e_.is_ok()
if s_even # true {
    intervals[s_index] = start; # start == 30
    s_index += 1 # s_index == 3 **HERE modification breaks below statement**
}
if e_even && s_index != e_index # true {
    e_index -= 1 # e_index == 1
    intervals[e_index] = end; # end == 40 
}
# result = [10, 40, 30, 50]
```

##### `s_index != e_index out of statement`

```
# [10, 20, 40, 50] + [30, 40], expected result = [10, 20, 30, 50]
# s_index = 2, e_index = 2, s.is_err(), e_.is_ok()
let s_not_equals_e = s_index != e_index;
if s_even # true {
    intervals[s_index] = start; # start == 30
    s_index += 1 # s_index == 3
}
if e_even && s_not_equals_e # false {
    e_index -= 1 
    intervals[e_index] = end; 
}
# result = [10, 20, 30, 50]
```

Reply:
agreed and changed regarding naming scheme